{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#! /usr/bin/env python\n",
    "\n",
    "import sys\n",
    "\n",
    "#for argument in sys.argv:\n",
    "\t#print(argument)\n",
    "\n",
    "#sys.exit()\n",
    "\n",
    "usage = sys.argv[0] + \": genome.fasta features.gff\"\n",
    "\n",
    "if len(sys.argv)<3:\n",
    "\tprint(usage)\n",
    "\tsys.exit(\"\\nThis script requires both a FASTA file and a GFF file\\n\")\n",
    "\n",
    "genome=sys.argv[1]\n",
    "gff=sys.argv[2]\n",
    "\n",
    "fsa_file=sys.argv[1]\n",
    "gff_file=sys.argv[2]\n",
    "fsa_in=open(fsa_file, 'r')\n",
    "gff_in=open(gff_file, 'r')\n",
    "\n",
    "#declare a variable for the genome sequence\n",
    "genome=''\n",
    "\n",
    "#initialize line counter\n",
    "first_line=0\n",
    "\n",
    "for line in fsa_in:\n",
    "    #print(str(first_line)+ ': '+ line)\n",
    "    line=line.rstrip(\"\\n\")\n",
    "    if first_line>0:\n",
    "        genome=genome+line\n",
    "        \n",
    "        #increment the line number\n",
    "    first_line+=1\n",
    "        \n",
    "fsa_in.close()\n",
    "\n",
    "cds=''\n",
    "trna=''\n",
    "rrna=''\n",
    "intron=''\n",
    "misc=''\n",
    "repeats=''\n",
    "\n",
    "\n",
    "\n",
    "#read in gff file\n",
    "\n",
    "for line in gff_in:\n",
    "    \n",
    "    line=line.rstrip(\"\\n\")\n",
    "    \n",
    "    types=line.split('; type')\n",
    "    other_type=types[len(types)-1]\n",
    "    #print(other_type)\n",
    "    \n",
    "    fields=line.split('\\t')\n",
    "    type=fields[2]\n",
    "    start=int(fields[3])\n",
    "    end=int(fields[4])\n",
    "    \n",
    "    #print(type,\"\\t\",start,\"\\t\",stop)\n",
    "    \n",
    "    #extract this feature from the genome\n",
    "    fragment=genome[start-1:end]\n",
    "    \n",
    "    if type=='CDS':\n",
    "        cds+=fragment\n",
    "    if type=='intron':\n",
    "        intron+=fragment\n",
    "    if type=='misc_feature':\n",
    "        misc+=fragment\n",
    "    if type=='repeat_region':\n",
    "        repeats+=fragment\n",
    "    if type=='rRNA':\n",
    "        rrna+=fragment\n",
    "    if type=='tRNA':\n",
    "        trna+=fragment\n",
    "#collate feature types and lengths\n",
    "\n",
    "#cds\n",
    "total_cds=len(cds)\n",
    "cds_gc=cds.count(\"G\")+cds.count(\"C\")\n",
    "#intron\n",
    "total_intron=len(intron)\n",
    "intron_gc=intron.count(\"G\")+intron.count(\"C\")\n",
    "#misc_feature\n",
    "total_misc=len(misc)\n",
    "misc_gc=misc.count(\"G\")+misc.count(\"C\")\n",
    "#rRNA\n",
    "total_rrna=len(rrna)\n",
    "rrna_gc=rrna.count(\"G\")+rrna.count(\"C\")\n",
    "#repeat_region\n",
    "total_repeats=len(repeats)\n",
    "repeats_gc=repeats.count(\"G\")+repeats.count(\"C\")\n",
    "#tRNA\n",
    "total_trna=len(trna)\n",
    "trna_gc=trna.count(\"G\")+trna.count(\"C\")\n",
    "#total_fragment\n",
    "total_fragment=len(fragment)\n",
    "\n",
    "    \n",
    "#print the output    \n",
    "print(\"exon\",\"\\t\",total_cds,\"\\t\",round(total_cds/len(genome)*100,1),\"\\t\",round((cds_gc/total_cds)*100,2))\n",
    "print(\"intron\",\"\\t\",total_intron,\"\\t\",round(total_intron/len(genome)*100,1),\"\\t\",round((intron_gc/total_intron)*100,2))\n",
    "print(\"misc_feature\",\"\\t\",total_misc,\"\\t\",round(total_misc/len(genome)*100,1),\"\\t\",round((misc_gc/total_misc)*100,2))\n",
    "print(\"rRNA\",\"\\t\",total_rrna,\"\\t\",round(total_rrna/len(genome)*100,1),\"\\t\",round((rrna_gc/total_rrna)*100,2))\n",
    "print(\"repeat_region\",\"\\t\",total_repeats,\"\\t\",round(total_repeats/len(genome)*100,1),\"\\t\",round((repeats_gc/total_repeats)*100,2))\n",
    "print(\"tRNA\",\"\\t\",total_trna,\"\\t\",round(total_trna/len(genome)*100,1),\"\\t\",round((trna_gc/total_trna)*100,2))\n",
    "\n",
    "    \n",
    "    #print(len(fragment), g_count, c_count)    \n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "#close the gff file\n",
    "gff_in.close()\n",
    "\n",
    "sys.exit()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
